src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs

```diff
@@ -2,168 +2,275 @@
+                    break;
+                case TypeCode.UInt16:
+                    if (reader.TryGetUInt32(out uint uint16) && JsonHelpers.IsInRangeInclusive(uint16, ushort.MinValue, ushort.MaxValue))
                     {
-                        if (reader.TryGetUInt64(out ulong uint64))
-                        {
-                            return Unsafe.As<ulong, T>(ref uint64);
-                        }
-                        break;
+                        ushort ushortValue = (ushort)uint16;
+                        return Unsafe.As<ushort, T>(ref ushortValue);
                     }
+                    break;
             }
 
             ThrowHelper.ThrowJsonException();
             return default;
         }
 
+        private static bool IsValidIdentifier(string value)
+        {
+            // Trying to do this check efficiently. When an enum is converted to
+            // string the underlying value is given if it can't find a matching
+            // identifier (or identifiers in the case of flags).
+            //
+            // The underlying value will be given back with a digit (e.g. 0-9) possibly
+            // preceded by a negative sign. Identifiers have to start with a letter
+            // so we'll just pick the first valid one and check for a negative sign
+            // if needed.
+            return (value[0] >= 'A' &&
+                (s_negativeSign == null || !value.StartsWith(s_negativeSign)));
+        }
+
         public override void Write(Utf8JsonWriter writer, T value, JsonSerializerOptions options)
         {
-            if (TreatAsString)
+            // If strings are allowed, attempt to write it out as a string value
+            if (_converterOptions.HasFlag(EnumConverterOptions.AllowStrings))
             {
-                writer.WriteStringValue(value.ToString());
+                string original = value.ToString();
+                if (_nameCache != null && _nameCache.TryGetValue(original, out string transformed))
+                {
+                    writer.WriteStringValue(transformed);
+                    return;
+                }
+
+                if (IsValidIdentifier(original))
+                {
+                    transformed = _namingPolicy.ConvertName(original);
+                    writer.WriteStringValue(transformed);
+                    if (_nameCache != null)
+                    {
+                        _nameCache.Add(original, transformed);
+                    }
+                    return;
+                }
             }
-            else if (s_enumTypeCode == TypeCode.UInt64)
+
+            if (!_converterOptions.HasFlag(EnumConverterOptions.AllowNumbers))
             {
-                // Use the ulong converter to prevent conversion into a signed\long value.
-                ulong ulongValue = Convert.ToUInt64(value);
-                writer.WriteNumberValue(ulongValue);
+                ThrowHelper.ThrowFormatException();
```

Comment:
This should be throwing JsonException unless you want a hard fail (end user bug). We no longer catch and re-throw `FormatException` as `JsonException`. Also applies to other cases here.

Reply:
Missed that change from yesterday.